include/aws/mqtt/private/client_impl.h

```diff
@@ -85,77 +88,102 @@ struct aws_mqtt_client_connection {
 
     struct aws_mqtt_client *client;
 
-    /* The host information */
+    /* Channel handler information */
+    struct aws_channel_handler handler;
+    struct aws_channel_slot *slot;
+
+    /* The host information, changed by user when state is AWS_MQTT_CLIENT_STATE_DISCONNECTED */
     struct aws_string *host_name;
     uint16_t port;
     struct aws_tls_connection_options tls_options;
     struct aws_socket_options socket_options;
 
+    /* Connect parameters */
+    struct aws_byte_buf client_id;
+    bool clean_session;
+    uint16_t keep_alive_time_secs;
+    uint64_t request_timeout_ns;
+    struct aws_string *username;
+    struct aws_string *password;
+    struct {
+        struct aws_byte_buf topic;
+        enum aws_mqtt_qos qos;
+        bool retain;
+        struct aws_byte_buf payload;
+    } will;
+    struct {
+        uint64_t current;      /* seconds */
+        uint64_t min;          /* seconds */
+        uint64_t max;          /* seconds */
+        uint64_t next_attempt; /* milliseconds */
+    } reconnect_timeouts;
     /* User connection callbacks */
     aws_mqtt_client_on_connection_complete_fn *on_connection_complete;
```

Comment:
These get set by the connect call, what if you connect while a connection/disconnection is in flight? Wouldn't these possibly produce a race? Should they be in the synced data?

Reply:
aws_mqtt_client_connection_connect only be called when state is AWS_MQTT_CLIENT_STATE_DISCONNECTED. So if the state is AWS_MQTT_CLIENT_STATE_DISCONNECTED, there will not be connection/disconnection in flight. BTW the other connection configuration set from seperate functions will only be called as CONNECTED or DISCONNECTED, so, they should be safe.